### DISCUSSION TITLE
HELP! How to delete SIMILAR (not EQUAL) frames in a GIF?

### DISCUSSION THREAD

**Discussion by Farivas** [ImageMagick/ImageMagick#1935] [General]
I have a project in which I need a lot of animated GIFs taken from animated series (cartoons and anime). Animated series usually have 24 fps, but not all frames are different, often there are frames that are repeated 2 or 3 times in a row.

My goal is to use ImageMagick to automate the removal of repeated frames.

I know that you will recommend me to use the next line of code (Source: http://www.imagemagick.org/Usage/anim_opt/#removedups ):

`convert input.gif -coalesce -layers RemoveDups output.gif`

However, this code is **limited** to frames that are **EXACTLY** the same, and the images I extract from the video files displayed have very slight JPEG errors ("noises") that cause the repeated frames to have a very slight difference between their pixels (that is, they are very **SIMILAR, but NOT EXACTLY THE SAME**).

Explanation of the attachments:
**S1_ORIG_A.gif**: (the one above) Is the GIF generated using ALL the frames extracted from the video. It weighs **6 Mb**.
**S1_OPT_A.gif**: (the one below) Is the GIF generated by deleting (manually) the repeated frames.  It weighs **3 Mb**, **half of the previous one**.

![S1_ORIG_A](https://user-images.githubusercontent.com/64302660/80274477-46446c00-86a0-11ea-9deb-79e163ca330e.gif)
![S1_OPT_A](https://user-images.githubusercontent.com/64302660/80274481-4cd2e380-86a0-11ea-955f-3cd4e4875d60.gif)

**Comment by snibgo** [comment]
I would do it in a script, something like this:

1. Extract all the frames into individual files.

2. Loop through the frame files, comparing each frame with the next, eg with "-metric RMSE". Where the score is below a certain threshold, consider the frames to be equal, so delete the first one.

3. From the remaining frames files, rebuild a GIF.

**Comment by Farivas** [comment]
I've been studying ImageMagick, but since I'm not a programmer (I barely know the basics), I have a slightly clearer idea of what I want to do.

1º Let the frame comparison method be the NCC method 
(https://imagemagick.org/script/compare.php), I've been trying it out and this (for the moment) to be the best method.
2º To compare a certain frame with its consecutive ones and when NCC>0.999 (if it fulfills, it replaces that frame), and if it doesn't fulfill, go to that frame that was not "similar" to the initial frame and to make the comparison with its consecutive ones again.
3º Finally, I could go through the RemoveDups method again (with the new generated gif) to directly remove those frames since they are exactly the same (in theory).

My flowchart is as follows ("n" is the number of rames of the gif minus 1; for example, if my gif has 5 frames, then n=5-1=4):

![Untitled Diagram (2)](https://user-images.githubusercontent.com/64302660/80406419-aba78100-8889-11ea-8b71-27e49b362906.png)

The problem is that I don't know how to implement it on the command line, as I said before, I'm not a programmer, I do other things.

**Reply by fmw42** [reply]
There is no simple command line that will do that. It will require scripting, which is OS dependent. What is your ImageMagick version and platform/OS?

**Reply by Farivas** [reply]
My OS: Windows 10
ImageMagick version: 7.0.10-7-portable-Q16-x64

**Reply by fmw42** [reply]
Sorry, I am not a Windows user. So cannot help. Perhaps one of the other Windows users may be able offer more help.

**Comment by vince-styling** [comment]
hi, have you found any solution to determinate if two noise frames are equals? I view you second gif frame by frame, notice since second frame there are striped off the identical pixels then before, just want to know how to do this

**Comment by snibgo** [comment]
@vince-styling: IM provides facilities to extract each frame from a GIF, and to compare any two frames to determine if they are similar (so one can be deleted by a shell command), and to combine the remaining files into a GIF. For example:
```
magick in.gif -layers coalesce +adjoin abc_%03d.miff

magick compare -metric RMSE abc_03.miff abc_05.miff NULL:

magick abc_*.miff -layers optimize out.gif
```
Those three operations can be used in a shell script to do the task.

**Reply by vince-styling** [reply]
yes, IM achieved a great optimizations for gif, I paid days on their course about optimization, looking for some source code to apply in my application

**Comment by God-damnit-all** [comment]
Were you ever able to figure this out? I'm interested in the result.
